File: apps/gmail/python/gmail_chatty_threads.py
# Copyright 2025 CyberWeb Consulting LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# Obtain a copy of the License at: apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# FILE:    gmail_chatty_threads.py
# POST:    dev.to/googleworkspace/...

from __future__ import print_function
import os

from google.auth.transport.requests import Request
from google_auth_oauthlib.flow import InstalledAppFlow as InstAppFlow
from googleapiclient import discovery
from google.oauth2.credentials import Credentials

# check credentials from locally-stored OAuth2 tokens file; either
# refresh expired tokens or run flow to get new pair & create API client
creds = None
SCOPES = 'https://www.googleapis.com/auth/gmail.metadata'
TOKENS = 'storage.json'  # where to store access & refresh tokens
if os.path.exists(TOKENS):
    creds = Credentials.from_authorized_user_file(TOKENS)
if not (creds and creds.valid):
    if creds and creds.expired and creds.refresh_token:
        creds.refresh(Request())
    else:
        flow: InstAppFlow = InstAppFlow.from_client_secrets_file(
                'client_secret.json', SCOPES)
        creds = flow.run_local_server()
with open(TOKENS, 'w') as token:
    token.write(creds.to_json())
GMAIL = discovery.build('gmail', 'v1', credentials=creds)

# scan your threads and individual msgs in each
threads = GMAIL.users().threads().list(userId='me').execute().get('threads', [])
for thread in threads:
    # process msgs for a thread
    tdata = GMAIL.users().threads().get(userId='me', id=thread['id'], format='metadata').execute()
    nmsgs = len(tdata['messages'])

    if nmsgs > 2:    # dump if <3 msgs in thread
        msg = tdata['messages'][0]['payload']
        subject = ''
        for header in msg['headers']:
            if header['name'] == 'Subject':
                subject = header['value']
                break
        if subject:  # dump if no Subject line
            print('-% 3d msgs: %s' % (nmsgs, subject if len(subject)<45 else '%s...' % subject[:42],))

File: apps/gmail/nodejs/gmail_chatty_threads.mjs
// Copyright 2025 CyberWeb Consulting LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// Obtain a copy of the License at: apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// FILE:    gmail_chatty_threads.mjs
// POST:    https://bit.ly/4ljSH2W

import fs from 'node:fs/promises';
import path from 'node:path';
import process from 'node:process';
import { authenticate } from '@google-cloud/local-auth';
import { google } from 'googleapis';

const CREDENTIALS_PATH = path.join(process.cwd(), 'client_secret.json');
const TOKEN_STORE_PATH = path.join(process.cwd(), 'storage.json');
const SCOPES = ['https://www.googleapis.com/auth/gmail.metadata'];


/**
 * Load any saved credentials or null
 *
 * @return {(Promise<JSONClient|null>)}
 */
async function loadSavedCredentialsIfExist() {
  try {
    const content = await fs.readFile(TOKEN_STORE_PATH);
    const credentials = JSON.parse(content);
    return google.auth.fromJSON(credentials);
  } catch (err) {
    return null;
  }
}

/**
 * Save credentials
 *
 * @param {JSONClient} client credentials
 * @return {Promise<void>}
 */
async function saveCredentials(client) {
  const content = await fs.readFile(CREDENTIALS_PATH);
  const keys = JSON.parse(content);
  const key = keys.installed || keys.web;
  const payload = JSON.stringify({
    type: 'authorized_user',
    client_id: key.client_id,
    client_secret: key.client_secret,
    refresh_token: client.credentials.refresh_token,
    access_token: client.credentials.access_token,
    token_expiry: client.credentials.token_expiry,
    scopes: client.credentials.scopes,
  });
  await fs.writeFile(TOKEN_STORE_PATH, payload);
}

/**
 * Authorize user credentials, authenticating user first if necessary
 *
 * @return {JSONClient} client credentials
 */
async function authorize() {
  var client = await loadSavedCredentialsIfExist();
  if (client) return client;
  client = await authenticate({
    scopes: SCOPES,
    keyfilePath: CREDENTIALS_PATH,
  });
  if (client.credentials) await saveCredentials(client);
  return client;
}

/**
 * Scan your threads and individual msgs in each
 *
 * @param {JSONClient} authClient (client credentials)
 */
async function gmailThreads(authClient) {
  const GMAIL = google.gmail({ version: 'v1', auth: authClient });
  var res = await GMAIL.users.threads.list({ userId: 'me' });
  const threads = res.data.threads;
  if (!threads || threads.length === 0) return;

  for (const thread of threads) {
    var res = await GMAIL.users.threads.get({
      userId: 'me',
      id: thread.id,
      format: 'metadata'
    });
    var tdata = res.data;
    var nmsgs = tdata.messages.length;

    if (nmsgs > 2) {    // dump if <3 msgs in thread
      var msg = tdata.messages[0].payload;
      var subject = '';
      for (const header of msg.headers) {
        if (header.name == 'Subject') {
          var subject = header.value;
          break
        }
      }
      if (subject) {  // dump if no Subject line
        console.log(`-${String(nmsgs).padStart(3, ' ')} msgs: ` +
          `${(subject.length < 45) ? subject : (subject.slice(0, 42)+'...')}`);
      }
    }
  }
}

authorize().then(gmailThreads).catch(console.error);
